aptos-move/framework/aptos-token/sources/listing.move

```diff
@@ -0,0 +1,245 @@
+    public fun create_list_under_user_account<CoinType>(
+        owner: &signer,
+        token_id: TokenId,
+        amount: u64,
+        min_price: u64,
+        instant_sale: bool,
+        expiration_sec: u64
+    ): ID acquires ListingRecords {
+        let owner_addr = signer::address_of(owner);
+        let record = create_list<CoinType>(owner, token_id, amount, min_price, instant_sale, expiration_sec);
+        initialize_listing_records<CoinType>(owner);
+        let records = borrow_global_mut<ListingRecords<CoinType>>(owner_addr);
+
+        let id = create_listing_id(owner);
+        // add a new record to the listing
+        table::add(&mut records.records, id, record);
+        id
+    }
+
+    /// creator uses this function to directly list token for sale under their own accounts
+    public entry fun direct_listing<CoinType>(
+        owner: &signer,
+        creator: address,
+        collection_name: String,
+        token_name: String,
+        property_version: u64,
+        amount: u64,
+        min_price: u64,
+        instant_sale: bool, // indicate if this listing is for sale or for auction
+        expiration_sec: u64
+    ) acquires ListingRecords {
+        let token_id = token::create_token_id_raw(creator, collection_name, token_name, property_version);
+        create_list_under_user_account<CoinType>(owner, token_id, amount, min_price, instant_sale, expiration_sec);
+    }
+
+    /// remove a listing for the direct listing records
+    public entry fun cancel_direct_list<CoinType>(
+        owner: &signer,
+        listing_id_creation_number: u64
+    ) acquires ListingRecords {
+        let listing_id = guid::create_id(signer::address_of(owner), listing_id_creation_number);
+        let owner_addr = signer::address_of(owner);
+        let records = &mut borrow_global_mut<ListingRecords<CoinType>>(owner_addr).records;
+        assert!(table::contains(records, listing_id), ELISTING_NOT_EXIST);
+        table::remove(records, listing_id);
+    }
+
+    /// internal function for assigned a global unique id for a listing
+    fun create_listing_id(owner: &signer): ID {
+        let gid = guid::create(owner);
+        guid::id(&gid)
+    }
+
+    public fun buy_internal<CoinType>(coin_owner: &signer, listing: Listing<CoinType>) {
+        assert!(listing.instant_sale, EINVALID_BUY_NOT_INSTANT_SALE);
+        let coin_owner_address = signer::address_of(coin_owner);
+
+        let total_amount = listing.min_price * listing.amount;
+        assert!(coin::balance<CoinType>(coin_owner_address) >= total_amount, EBUYER_NOT_HAVING_ENOUGH_COINS);
+
+        let token = token::withdraw_with_event_internal(listing.owner, listing.token_id, listing.amount);
+        coin::transfer<CoinType>(coin_owner, listing.owner, total_amount);
+        token::direct_deposit(signer::address_of(coin_owner), token);
+    }
+
+    public entry fun buy<CoinType>(coin_owner: &signer, token_owner: &signer, listing_id: u64) acquires ListingRecords {
```

Comment:
this doesn't require token_owner's signer. When listing token, token owner already signed the listing.

Reply:
maybe rename the listing_id to id_creation_number